src/msgraph_core/tasks/page_iterator.py

```diff
@@ -0,0 +1,230 @@
+     constructor_callable: Optional[Callable] = None): Initializes a new instance of 
+     the PageIterator class.
+    """
+
+    def __init__(
+        self,
+        response: Union[T, list, object],
+        request_adapter: HttpxRequestAdapter,
+        constructor_callable: Optional[Callable] = None
+    ):
+        self.request_adapter = request_adapter
+
+        if isinstance(response, Parsable) and not constructor_callable:
+            constructor_callable = getattr(type(response), 'create_from_discriminator_value')
+        elif constructor_callable is None:
+            constructor_callable = PageResult.create_from_discriminator_value
+        self.pause_index = 0
+        self.headers: HeadersCollection = HeadersCollection()
+        self.request_options = []  # type: ignore
+        self.current_page = self.convert_to_page(response)
+        self.object_type = self.current_page.value[
+            0].__class__.__name__ if self.current_page.value else None
+        page = self.current_page
+        self._next_link = response.get('@odata.nextLink', '') if isinstance(
+            response, dict
+        ) else getattr(response, '@odata.nextLink', '')
+        self._delta_link = response.get('@odata.deltaLink', '') if isinstance(
+            response, dict
+        ) else getattr(response, '@odata.deltaLink', '')
+
+        if page is not None:
+            self.current_page = page
+            self.has_next = bool(page.odata_next_link)
+
+    def set_headers(self, headers: dict) -> None:
+        """
+        Sets the headers for the HTTP requests.
+        This method takes a dictionary of headers and adds them to the
+        existing headers.
+        Args:
+            headers (dict): A dictionary of headers to add. The keys are the 
+            header names and the values are the header values.
+        """
+        self.headers.add_all(**headers)
+
+    @property
+    def delta_link(self):
+        return self._delta_link
+
+    @property
+    def next_link(self):
+        return self._next_link
+
+    def set_request_options(self, request_options: list) -> None:
+        """
+        Sets the request options for the HTTP requests.
+        Args:
+            request_options (list): The request options to set.
+        """
+        self.request_options = request_options
+
+    async def iterate(self, callback: Callable) -> None:
+        """
+        Iterates over the pages and applies a callback function to each item.
+        The iteration stops when there are no more pages or the callback
+         function returns False.
+        Args:
+            callback (Callable): The function to apply to each item. 
+            It should take one argument (the item) and return a boolean.
+        """
+        while True:
+            keep_iterating = self.enumerate(callback)
+            if not keep_iterating:
+                return
+            next_page = await self.next()
+            if not next_page:
```

Comment:
[question] Should we also set `has_next` to `False` here?

Reply:
We get has_next from next method, which uses odata_next_link, it will give a false if we are at end of the pagedCollection without us needing to update it here